### PR TITLE
Support `<gazebo_ros>` in `package.xml` exports

### PR DESCRIPTION
**Pull request by azeey**
# 🎉 New feature

## Summary
This copies the implementation from `gazebo_ros_paths.py` to provide a way for packages to set resource paths from `package.xml`.

```
e.g.  <export>
          <gazebo_ros gazebo_model_path="${prefix}/models"/>
          <gazebo_ros gazebo_media_path="${prefix}/models"/>
      </export>
```

The value of `gazebo_model_path` and `gazebo_media_path` is appended to `GZ_SIM_RESOURCE_PATH` The value of `plugin_path` appended to `GZ_SIM_SYSTEM_PLUGIN_PATH`.

This will make transition from Gazebo Classic easier.

## Checklist
- [ ] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

**Review by ahcorde — APPROVED** [review]
it might be worth it to add a tutorial about this or small doc about this. 

Do you have plans to forward port this to Iron and Rolling?

**Comment by azeey** [comment]
> it might be worth it to add a tutorial about this or small doc about this.

Good point!

> Do you have plans to forward port this to Iron and Rolling?

Yes, I'll cherry-pick this to Iron and Rolling. I was going to start on Rolling first, but I wanted to make sure the `humble` branch got the `IGN_*` variables modified as well.

File: ros_gz_sim/launch/gz_sim.launch.py.in
Comment (j-rivero) [inline ros_gz_sim/launch/gz_sim.launch.py.in:64]: I think that one of the most frustrating issue for newcomers and people dealing with Gazebo + ROS integration is to badly configure the asset directories assuming random paths that are not the ones really processed here. I would like to use the PR to improve this situation.

How about adding a bit of sanity checks before the `.append(` calls in this file that can emit warning/errors for the user to check/fix, some ideas:
  1. The `xml_path` in a vast majority of the cases needs to be an existing non empty directory.
  2. For the `gazebo_model_path`, we can probably scan for the expected subdirectory structure with at least one .sdf and one .config.
  3. For `gazebo_plugin_path` we can probably expect that the user hosts at least one .so or one binary file.

File: ros_gz_sim/launch/gz_sim.launch.py.in
Comment (j-rivero) [inline ros_gz_sim/launch/gz_sim.launch.py.in]: Are `model_path` and `media_path` being used exactly the same? If so we can simplify the code just by return one that merges both (resource_path) from `GazeboRosPaths`.

**Review by j-rivero — COMMENTED** [review]
Just commenting here what we spoke in the weekly meeting: let's try to create a tutorial that documents all the options of injecting the paths available and point from this code to the user to specify that this is not the easiest way or the proffered way with a pointer to the tutorial.

**Comment by DLu** [comment]
Are we waiting on the Tutorial to get this merged? 

I have a [URDF/Gazebo tutorial waiting to be merged](https://github.com/ros2/ros2_documentation/pull/3952) and there is a section ("Side note: Configuring meshes") that covers this area. The tutorial is written for for Gazebo Classique, but as linked above, I'm trying to get the Gazebo New Coke version to work too.

File: ros_gz_sim/launch/gz_sim.launch.py.in
Comment (azeey) [reply ros_gz_sim/launch/gz_sim.launch.py.in]: I believe model and media paths are combined into resource path now, so I've simplified the code as you suggested.

File: ros_gz_sim/launch/gz_sim.launch.py.in
Comment (azeey) [reply ros_gz_sim/launch/gz_sim.launch.py.in:64]: I looked at doing this but the way this works, it's actually looking at all the packages in the system, not just the individual package the user might be interested in. And would have to do the checks for all packages every time this file is launched. Aside from the performance implications, a user launching this file would see errors/warnings coming from other packages they have no control over. 

An alternative is to create a tool that can be used to search for a particular model/resource by it's URI. For example, if I wanted to check that the URI `model://my_awesome_pkg/models/cool_robot` is valid, I would run this tool  and it would search all exported model paths and give me some diagnostic messages if the model couldn't be found.

**Comment by azeey** [comment]
@ahcorde , @DLu I've added a section to the README that explains how to use this. Let me know your thoughts. This could be expanded into a full tutorial on gazebosim.org/docs.

File: ros_gz_sim/launch/gz_sim.launch.py.in
Comment (azeey) [reply ros_gz_sim/launch/gz_sim.launch.py.in:64]: @j-rivero any thoughts? I'd like to unblock https://github.com/ros/urdf_sim_tutorial/issues/16 soon if possible.

**Comment by DLu** [comment]
Tip the hat

File: ros_gz_sim/launch/gz_sim.launch.py.in
Comment (j-rivero) [reply ros_gz_sim/launch/gz_sim.launch.py.in:64]: > I looked at doing this but the way this works, it's actually looking at all the packages in the system, not just the individual package the user might be interested in.

I was thinking on limiting the directories to inspect to the ones added via:
```xml
<export>
   <gazebo_ros gazebo_model_path="${prefix}/models"/>
   <gazebo_ros gazebo_media_path="${prefix}/media"/>
   <gazebo_ros plugin_path="${prefix}/plugins"/>
</export>
```
Why do we need to check all the packages in the system?

> And would have to do the checks for all packages every time this file is launched. Aside from the performance implications,

I agree that there is a performance implication that can be bad.

> An alternative is to create a tool that can be used to search for a particular model/resource by it's URI. For example, if I wanted to check that the URI model://my_awesome_pkg/models/cool_robot is valid, I would run this tool and it would search all exported model paths and give me some diagnostic messages if the model couldn't be found.

That would be really useful indeed. Another approach would be to enable a launch argument or environment flag that enable this debug behavior.

Given the situation, I'm not blocking the PR, thought it was easier to add the checks.

**Review by j-rivero — APPROVED** [review]

File: ros_gz_sim/launch/gz_sim.launch.py.in
Comment (azeey) [reply ros_gz_sim/launch/gz_sim.launch.py.in:64]: The code `for package_name in get_package_names():` iterates through all the packages available. Then we'd need to check if any package that exports `gazebo_ros` meets the sanity requirements you described.

**Comment by ahcorde** [comment]
@azeey do you have any plan to forward port this to iron and rolling?

**Comment by rakeshv24** [comment]
@azeey is this feature included in the jazzy release yet?

**Comment by azeey** [comment]
Ah, this fell through the cracks. I've started the forward port process: https://github.com/gazebosim/ros_gz/pull/564

**Comment by rakeshv24** [comment]
Thanks!

**Pull request merged**
